File: repositorio_incidentes.py
import math
import pandas as pd

from supabase_client import get_client

TABELA_INCIDENTES = "incidentes"
TABELA_DIAS_NAO_UTEIS = "dias_nao_uteis"

_MAPA_COLUNAS = {
    "Number": "number",
    "Service": "service",
    "Opened": "opened",
    "Resolved": "resolved",
    "Short description": "short_description",
    "State": "state",
    "Priority": "priority",
    "Assignment group": "assignment_group",
    "ICT Service": "ict_service",
    "Assigned to": "assigned_to",
    "Close code": "close_code",
    "Close notes": "close_notes",
    "Reopen count": "reopen_count",
    "Empresa": "empresa",
    "Macro": "macro",
    "Categoria": "categoria",
    "SubCategoria": "subcategoria",
    "Descricao_Tratada": "descricao_tratada",
    "SLA - Dias (8 h)": "sla_dias",
    "TMA - Dias corridos": "tma_dias",
}

_COLUNAS_INVERTIDAS = {v: k for k, v in _MAPA_COLUNAS.items()}


def _limpar_valor(v):
    if v is None:
        return None
    if isinstance(v, float) and math.isnan(v):
        return None
    if isinstance(v, pd.Timestamp):
        if pd.isna(v):
            return None
        return v.isoformat()
    try:
        if pd.isna(v):
            return None
    except (TypeError, ValueError):
        pass
    return v


def dataframe_para_registros(df: pd.DataFrame) -> list[dict]:
    """Converte o dataframe já processado (df_completo) em registros no
    formato da tabela `incidentes`, removendo duplicados de Number
    (mantendo o último, que costuma ser o dado mais atualizado)."""
    df = df.drop_duplicates(subset=["Number"], keep="last")

    registros = []
    for _, row in df.iterrows():
        registro = {}
        for col_origem, col_destino in _MAPA_COLUNAS.items():
            registro[col_destino] = _limpar_valor(row.get(col_origem)) if col_origem in df.columns else None
        registros.append(registro)
    return registros


def upsert_incidentes(df: pd.DataFrame, tamanho_lote: int = 500, client=None) -> int:
    """Insere incidentes novos e atualiza os existentes (upsert por `number`).
    Retorna a quantidade de registros enviados."""
    registros = dataframe_para_registros(df)
    if not registros:
        return 0

    client = client or get_client()
    total = 0
    for i in range(0, len(registros), tamanho_lote):
        lote = registros[i:i + tamanho_lote]
        client.table(TABELA_INCIDENTES).upsert(lote, on_conflict="number").execute()
        total += len(lote)
    return total


def carregar_incidentes() -> pd.DataFrame:
    """Carrega todos os incidentes do Supabase, já convertidos para o
    formato de colunas usado pelo restante do app (df_completo)."""
    client = get_client()
    registros = []
    pagina = 1000
    inicio = 0
    while True:
        resp = (
            client.table(TABELA_INCIDENTES)
            .select("*")
            .range(inicio, inicio + pagina - 1)
            .execute()
        )
        dados = resp.data
        registros.extend(dados)
        if len(dados) < pagina:
            break
        inicio += pagina

    if not registros:
        colunas = list(_MAPA_COLUNAS.keys())
        return pd.DataFrame(columns=colunas)

    df = pd.DataFrame(registros)
    df = df.rename(columns=_COLUNAS_INVERTIDAS)
    df["Opened"] = pd.to_datetime(df["Opened"], errors="coerce", utc=True).dt.tz_localize(None)
    df["Resolved"] = pd.to_datetime(df["Resolved"], errors="coerce", utc=True).dt.tz_localize(None)
    return df


_CAMPOS_EDITAVEIS = {"Macro", "Categoria", "SubCategoria", "Descricao_Tratada"}


def atualizar_campos_incidente(number: str, campos: dict):
    """Atualiza campos pontuais (ex: Macro/Categoria/SubCategoria/Descricao_Tratada)
    de um incidente já existente, identificado pelo Number."""
    payload = {
        _MAPA_COLUNAS[campo]: _limpar_valor(valor)
        for campo, valor in campos.items()
        if campo in _CAMPOS_EDITAVEIS
    }
    if not payload:
        return
    client = get_client()
    client.table(TABELA_INCIDENTES).update(payload).eq("number", number).execute()


def carregar_dias_nao_uteis() -> pd.DataFrame:
    client = get_client()
    resp = client.table(TABELA_DIAS_NAO_UTEIS).select("*").order("data").execute()
    return pd.DataFrame(resp.data)


def adicionar_dia_nao_util(data_iso: str, motivo: str = ""):
    client = get_client()
    client.table(TABELA_DIAS_NAO_UTEIS).upsert(
        {"data": data_iso, "motivo": motivo}, on_conflict="data"
    ).execute()


def remover_dia_nao_util(dia_id: int):
    client = get_client()
    client.table(TABELA_DIAS_NAO_UTEIS).delete().eq("id", dia_id).execute()

File: web/lib/repositorio-incidentes.ts
import { getClient } from "./supabase";
import type { Incidente, DiaNaoUtil } from "./types";
import { CAMPOS_EDITAVEIS } from "./types";

const TABELA_INCIDENTES = "incidentes";
const TABELA_DIAS_NAO_UTEIS = "dias_nao_uteis";

const MAPA_COLUNAS: Record<keyof Incidente, string> = {
  Number: "number",
  Service: "service",
  Opened: "opened",
  Resolved: "resolved",
  "Short description": "short_description",
  State: "state",
  Priority: "priority",
  "Assignment group": "assignment_group",
  "ICT Service": "ict_service",
  "Assigned to": "assigned_to",
  "Close code": "close_code",
  "Close notes": "close_notes",
  "Reopen count": "reopen_count",
  Empresa: "empresa",
  Macro: "macro",
  Categoria: "categoria",
  SubCategoria: "subcategoria",
  Descricao_Tratada: "descricao_tratada",
  "SLA - Dias (8 h)": "sla_dias",
  "TMA - Dias corridos": "tma_dias",
};

const COLUNAS_INVERTIDAS = Object.fromEntries(
  Object.entries(MAPA_COLUNAS).map(([k, v]) => [v, k])
) as Record<string, string>;

function incidenteParaRegistro(inc: Incidente): Record<string, unknown> {
  const registro: Record<string, unknown> = {};
  for (const [colOrigem, colDestino] of Object.entries(MAPA_COLUNAS)) {
    registro[colDestino] = (inc as unknown as Record<string, unknown>)[colOrigem] ?? null;
  }
  return registro;
}

export async function upsertIncidentes(incidentes: Incidente[], tamanhoLote = 500): Promise<number> {
  // dedup por Number, mantendo o último
  const porNumber = new Map<string, Incidente>();
  for (const inc of incidentes) porNumber.set(inc.Number, inc);
  const registros = Array.from(porNumber.values()).map(incidenteParaRegistro);

  if (registros.length === 0) return 0;

  const client = getClient();
  let total = 0;
  for (let i = 0; i < registros.length; i += tamanhoLote) {
    const lote = registros.slice(i, i + tamanhoLote);
    const { error } = await client.from(TABELA_INCIDENTES).upsert(lote, { onConflict: "number" });
    if (error) throw new Error(`Erro ao salvar incidentes: ${error.message}`);
    total += lote.length;
  }
  return total;
}

export async function carregarIncidentes(): Promise<Incidente[]> {
  const client = getClient();
  const pagina = 1000;
  let inicio = 0;
  const registros: Record<string, unknown>[] = [];

  for (;;) {
    const { data, error } = await client
      .from(TABELA_INCIDENTES)
      .select("*")
      .range(inicio, inicio + pagina - 1);
    if (error) throw new Error(`Erro ao carregar incidentes: ${error.message}`);
    registros.push(...(data ?? []));
    if (!data || data.length < pagina) break;
    inicio += pagina;
  }

  return registros.map((row): Incidente => {
    const incidente: Record<string, unknown> = {};
    for (const [colBanco, colOrigem] of Object.entries(COLUNAS_INVERTIDAS)) {
      incidente[colOrigem] = row[colBanco] ?? null;
    }
    return incidente as unknown as Incidente;
  });
}

export async function atualizarCamposIncidente(
  number: string,
  campos: Partial<Record<(typeof CAMPOS_EDITAVEIS)[number], string>>
): Promise<void> {
  const payload: Record<string, unknown> = {};
  for (const [campo, valor] of Object.entries(campos)) {
    if ((CAMPOS_EDITAVEIS as readonly string[]).includes(campo)) {
      payload[MAPA_COLUNAS[campo as keyof Incidente]] = valor ?? null;
    }
  }
  if (Object.keys(payload).length === 0) return;

  const client = getClient();
  const { error } = await client.from(TABELA_INCIDENTES).update(payload).eq("number", number);
  if (error) throw new Error(`Erro ao atualizar incidente: ${error.message}`);
}

export async function carregarDiasNaoUteis(): Promise<DiaNaoUtil[]> {
  const client = getClient();
  const { data, error } = await client.from(TABELA_DIAS_NAO_UTEIS).select("*").order("data");
  if (error) throw new Error(`Erro ao carregar dias não úteis: ${error.message}`);
  return (data ?? []) as unknown as DiaNaoUtil[];
}

export async function adicionarDiaNaoUtil(dataIso: string, motivo = ""): Promise<void> {
  const client = getClient();
  const { error } = await client
    .from(TABELA_DIAS_NAO_UTEIS)
    .upsert({ data: dataIso, motivo }, { onConflict: "data" });
  if (error) throw new Error(`Erro ao adicionar dia não útil: ${error.message}`);
}

export async function removerDiaNaoUtil(id: number): Promise<void> {
  const client = getClient();
  const { error } = await client.from(TABELA_DIAS_NAO_UTEIS).delete().eq("id", id);
  if (error) throw new Error(`Erro ao remover dia não útil: ${error.message}`);
}
